### PR TITLE
feat(decide): Move flag evaluation to read replica

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

Move flag evaluation to the decide replica.

Also adds an extensive test suite that _should_ help us catch when we're going to the main db when we shouldn't.

Django is pretty terrible when it comes to routing like this. We need to be explicit everywhere and be very careful. Here's a nasty bug I caught thanks to the test suite:

```
with transaction.atomic():
        with connections['replica'].cursor() as cursor:
            cursor.execute("SET LOCAL statement_timeout = %s", [timeout])
            yield cursor
```

The above is wrong, because the transaction.atomic() context manager is going to the main DB,  while the cursor below is going to the replica. This effectively means that the query to the replica is not a transaction. Booo!

---

So, I've covered all code paths in the new test class, and assert that the right number of queries make it to the replica. Have added this separately to the CI as well, since overriding settings doesn't work well when it comes to database setting - you need to set it at the env var level before tests.

In the future, whenever we make any db related changes to decide, should add the new route to this test suite.

We also have special handling for the write-path in decide, and whenever we go write, we subsequently read from the main db, not the replica, to ensure replica lag doesn't bork evaluation.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Lots of tests!

### CODE REVIEW

File: posthog/models/feature_flag/flag_matching.py [hunk 1]
```diff
@@ -323,13 +328,13 @@ def query_conditions(self) -> Dict[str, bool]:
...
-            with execute_with_timeout(FLAG_MATCHING_QUERY_TIMEOUT_MS * 2):
+            with execute_with_timeout(FLAG_MATCHING_QUERY_TIMEOUT_MS * 2, DATABASE_FOR_FLAG_MATCHING):
                 all_conditions: Dict = {}
                 team_id = self.feature_flags[0].team_id
-                person_query: QuerySet = Person.objects.filter(
+                person_query: QuerySet = Person.objects.using(DATABASE_FOR_FLAG_MATCHING).filter(
                     team_id=team_id, persondistinctid__distinct_id=self.distinct_id, persondistinctid__team_id=team_id
                 )
-                basic_group_query: QuerySet = Group.objects.filter(team_id=team_id)
+                basic_group_query: QuerySet = Group.objects.using(DATABASE_FOR_FLAG_MATCHING).filter(team_id=team_id)
```
Comment (neilkakkar) [inline posthog/models/feature_flag/flag_matching.py:337]: oh, forgot to test for the cohorts path, that will probably still fail 🤔

File: posthog/database_healthcheck.py [hunk 2]
```diff
@@ -43,8 +43,10 @@ def _get_timebucket(self) -> int:
         return round(time.time() / self.time_interval)
 
     def is_postgres_connected_check(self) -> bool:
+        from posthog.models.feature_flag.flag_matching import DATABASE_FOR_FLAG_MATCHING
```
Comment (hazzadous) [inline posthog/database_healthcheck.py:46]: Can this import be moved to the top level? Or do you end up with cyclic imports? If the latter, is it possible to refactor to have a non-cyclic dependency graph by moving DATABASE_FOR_FLAG_MATCHING to a separate common dependency file?

File: posthog/models/feature_flag/feature_flag.py [hunk 3]
```diff
@@ -298,7 +298,11 @@ def set_feature_flags_for_team_in_cache(
 
     serialized_flags = MinimalFeatureFlagSerializer(all_feature_flags, many=True).data
 
-    cache.set(f"team_feature_flags_{team_id}", json.dumps(serialized_flags), FIVE_DAYS)
+    try:
+        cache.set(f"team_feature_flags_{team_id}", json.dumps(serialized_flags), FIVE_DAYS)
+    except Exception:
+        # redis is unavailable
+        capture_exception()
```
Comment (hazzadous) [inline posthog/models/feature_flag/feature_flag.py:305]: Does this also log to std out already? If not then I'd add a log message here so we're not dependent on Sentry for introspection.

File: posthog/models/feature_flag/flag_matching.py [hunk 4]
```diff
@@ -44,6 +45,8 @@
     labelnames=["reason"],
 )
 
+DATABASE_FOR_FLAG_MATCHING = "default" if "decide" not in settings.READ_REPLICA_OPT_IN else "replica"
```
Comment (hazzadous) [inline posthog/models/feature_flag/flag_matching.py:48]: I'm guessing moving this to settings will remove and cyclic dependency issues if there were any?

**Review by hazzadous — APPROVED** [review]
Looks good to me

File: posthog/database_healthcheck.py (hunk 2 above)
Comment (neilkakkar) [reply posthog/database_healthcheck.py:46]: Ah, no cyclic imports, was testing something different, forgot to move it back up, will do!

Comment (neilkakkar) [reply posthog/database_healthcheck.py:46]: nvm, is cyclic, will move FLAG_MATCHING elsewhere

**Comment by neilkakkar** [comment]
Let's gooo! Should be safe to go to prod since env var controlled

**Pull request merged**